mesonbuild/modules/python.py

```diff
@@ -390,13 +414,18 @@ def sanity(self) -> bool:
             mlog.debug('Program stderr:\n')
             mlog.debug(stderr)
 
-        if info is not None and self._check_version(info['version']):
-            variables = info['variables']
-            info['suffix'] = variables.get('EXT_SUFFIX') or variables.get('SO') or variables.get('.so')
-            self.info = T.cast('PythonIntrospectionDict', info)
-            self.platlib = self._get_path('platlib')
-            self.purelib = self._get_path('purelib')
-            return True
+        if info is not None:
+            if 'error' in info:
+                assert isinstance(info['error'], str)
+                mlog.error('Python interpreter introspection failed: %s' % info['error'])
```

Comment:
Please use f-strings (or, when the complexity penalty is too great, the `str.format()` method).

The assert doesn't seem to be performing any debug sanity checks, so it can probably be dropped.

Reply:
The assert is to make sure the nobody screw up when editing the introspection script, but I can remove it if you want.

Reply:
Actually that may speak to a greater need: adding a test case that actually tries to trigger this code path and testing that it produces the correct error message.

On the other hand -- if it's not a string what will it be? It is loaded as json there are only so many options...

Reply:
I don't have strong opinions on the assert, but as a project standard we have purged percent formatting in favor of f-strings wherever it doesn't have major readability costs (now that we require a minimum python which supports f-strings) and the .format() method otherwise.

Reply:
> Actually that may speak to a greater need: adding a test case that actually tries to trigger this code path and testing that it produces the correct error message.

We would have to keep the tests in updated to make sure they cover all code paths in the introspection command.

> On the other hand -- if it's not a string what will it be? It is loaded as json there are only so many options...

Could be anything. I just put the assert there as a sanity check to avoid obscure error. Yes, this is me being preventive, but it's cheap, so...

> I don't have strong opinions on the assert, but as a project standard we have purged percent formatting in favor of f-strings wherever it doesn't have major readability costs (now that we require a minimum python which supports f-strings) and the .format() method otherwise.

Of course. I don't like nesting quotes in f-strings, so I went with `%` as that's what the neighboring code did, but I can definitely change.

Reply:
Nesting quotes? Do you mean like `f'{some string!r}'`  to guarantee that the printed content itself has quite characters?

Reply:
```python
f'something: {info["error"]}'
```

Reply:
Also it is true we do have some percent formatted strings still... I would like to get rid of all of them but I mainly drove the upgrade using pyupgrade which is pretty good but still heuristics based and thus not perfect. The vast majority were changed though. :D

Reply:
Anyway, I change to `str.format`, let me know if you'd like me to change to an f-string.